### ISSUE TITLE
[Bug] Merge Failed if there is a `.lock` file exists

### ISSUE THREAD

**Issue by Whirlwind** [libgit2/libgit2#5642]
### Reproduction steps
I use the `git_merge` to merge other branch into current branch. But I always get a error: `failed to lock file 'Podfile.lock' for writing:`. 
I try to use the `git merge` command to merge the branch, the git will merge success with conflict: The `Podfile` have conflict.
I check the `git_filebuf_open_withsize` function in `libgit2`, I get:
```

		/* create the locking path by appending ".lock" to the original */
		GIT_ERROR_CHECK_ALLOC_ADD(&alloc_len, path_len, GIT_FILELOCK_EXTLENGTH);
		file->path_lock = git__malloc(alloc_len);
		GIT_ERROR_CHECK_ALLOC(file->path_lock);

		memcpy(file->path_lock, file->path_original, path_len);
		memcpy(file->path_lock + path_len, GIT_FILELOCK_EXTENSION, GIT_FILELOCK_EXTLENGTH);
```
The `file->path_lock` will append the `.lock` to the original. And the original path is `Podfile`, so the `path_lock` is `Podfile.lock`. But the `Podfile.lock` exists!!! 

Is it a bug?

### Expected behavior

Merge success with a conflict file.

### Actual behavior

Get a error: `failed to lock file 'Podfile.lock' for writing:`
### Version of libgit2 (release number or SHA1)
v1.0.1
### Operating system(s) tested
macOS 10.15

**Comment by 0xdky** [comment]
> But the `Podfile.lock` exists!!!

That is the problem. You have a branch named `Podfile` and a ref update has failed with incomplete cleanup. You could rename the file (safer than deleting it) to `Podfile.lock.rm` and give it a try. If it works, you can delete `Podfile.lock.rm` and move on.

The way ref lock files work is by opening a file with ref name appended by `.lock` in `O_CREAT|O_EXCL` mode. If the file exists, it is an indication that the ref has been locked and a ref update might be in progress. To avoid clobbering the ref, you get a failure instead.

Using file system for creating locks will open up for stale locks due to improper cleanup and there is no easy work around for it without using file system provided locking which gets released when the process holding it is killed. Since not all file systems implement a standard compliant byte range locking or file locks, hence git uses a more primitive approach to implement locks using exclusive file create operation.

**Comment by Whirlwind** [comment]
I known the `Podfile` + `.lock` is a file lock, but in my project there are `Podfile` and `Podfile.lock` at same time.

I don't known why to write a file lock in same directory, not in the `.git` directory. Or use other extension instead of `.lock`.
The `.lock` is too simple!!

**Comment by ethomson** [comment]
> The .lock is too simple!!

Yep, this seems like a bug indeed.  I'm surprised that we hadn't seen it before.  Thanks for reporting it.
